### PR TITLE
Assignment5 - Chester Southwood

### PR DESCRIPTION
**Pull request by Chester-Southwood**

### CODE REVIEW

File: SecretSanta/src/SecretSanta.Domain/Services/PairingService.cs [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+        private ApplicationDbContext DbContext { get; }
+
+        public PairingService(ApplicationDbContext dbContext)
+        {
+            DbContext = dbContext ?? throw new ArgumentNullException(nameof(dbContext));
+        }
+
+        public async Task<bool> GeneratePairing(int groupId)
+        {
+            //id is primary key
```
Comment (efueng) [inline SecretSanta/src/SecretSanta.Domain/Services/PairingService.cs]: I stole these comments from you. Great reference! Should probably be removed for the final PR however

File: SecretSanta/test/SecretSanta.Api.Tests/Controllers/GiftControllerTests.cs [hunk 2]
```diff
@@ -22,29 +23,31 @@ public static void ConfigureAutoMapper(TestContext context)
...
                 {
                     gift
-                }
+                })
             };
-            var controller = new GiftsController(testService, Mapper.Instance);
+            GiftsController controller = new GiftsController(testService, Mapper.Instance);
 
-            var result = controller.GetGiftForUser(4) as OkObjectResult;
+            IActionResult result = await controller.GetGiftForUser(4);
```
Comment (efueng) [inline SecretSanta/test/SecretSanta.Api.Tests/Controllers/GiftControllerTests.cs]: Did you split `var result = await controller.GetGiftsForUser(4) as OkObjectResult;` into lines 45 and 46 for readability reasons?

File: SecretSanta/test/SecretSanta.Api.Tests/Controllers/GroupControllerTests.cs [hunk 3]
```diff
@@ -10,37 +10,39 @@
...
-                .Returns(new List<Group> { group1, group2 })
+                .Returns(Task.Run( () => new List<Group> { group1, group2 }))
                 .Verifiable();
 
 
-            var controller = new GroupsController(service.Object, Mapper.Instance);
+            GroupsController controller = new GroupsController(service.Object, Mapper.Instance);
 
-            var result = controller.Get() as OkObjectResult;
+            IActionResult result = await Task.Run(() => controller.Get());
```
Comment (efueng) [inline SecretSanta/test/SecretSanta.Api.Tests/Controllers/GroupControllerTests.cs]: In case you're not splitting `var result = controller.Get() as OkObjectResult;` just for readability reasons, I *think* the `as` conversion automatically extracts the `IActionResult` from the returned `Task` for you.

File: SecretSanta/test/SecretSanta.Api.Tests/Controllers/GroupControllerTests.cs [hunk 4]
```diff
@@ -49,130 +51,156 @@ public void GetAllGroups_ReturnsGroups()
...
+        public async Task CreateGroup_RequiresGroup()
         {
-            var service = new Mock<IGroupService>(MockBehavior.Strict);
-            var controller = new GroupsController(service.Object, Mapper.Instance);
+            Mock<IGroupService> service = new Mock<IGroupService>(MockBehavior.Strict);
+            GroupsController controller = new GroupsController(service.Object, Mapper.Instance);
 
-
-            var result = controller.Post(null) as BadRequestResult;
+            IActionResult task = await Task.Run(() => controller.Post(null));
```
Comment (efueng) [inline SecretSanta/test/SecretSanta.Api.Tests/Controllers/GroupControllerTests.cs]: Again, I don't think you need to split this up. Keeping it as `var result = controller.Post(null) as BadRequestResult;` will also let you avoid calling `Task.Run()`. I may have misunderstood what Kevin was saying, but I think we generally want to avoid calling `Task.Run()`. But I'm not totally sure about that, so if one of the instructors sees this and I'm wrong, PLEASE correct me.

File: SecretSanta/test/SecretSanta.Api.Tests/Controllers/GroupControllerTests.cs [hunk 5]
```diff
@@ -49,130 +51,156 @@ public void GetAllGroups_ReturnsGroups()
...
+        public async Task CreateGroup_ReturnsCreatedGroup()
         {
             var group = new GroupInputViewModel
             {
                 Name = "Group"
             };
             var service = new Mock<IGroupService>();
             service.Setup(x => x.AddGroup(It.Is<Group>(g => g.Name == group.Name)))
-                .Returns(new Group
+                .Returns(Task.Run(() => new Group
```
Comment (efueng) [inline SecretSanta/test/SecretSanta.Api.Tests/Controllers/GroupControllerTests.cs]: Using `ReturnsAsync` lets you avoid calling `Task.Run()`.
```csharp
service.Setup(x => x.GetById(2))
                .ReturnsAsync(new Group
                {
                    Id = 2,
                    Name = group.Name
                })
                .Verifiable();
```

File: SecretSanta/test/SecretSanta.Api.Tests/Controllers/GroupControllerTests.cs [hunk 6]
```diff
@@ -49,130 +51,156 @@ public void GetAllGroups_ReturnsGroups()
...
         {
-            var service = new Mock<IGroupService>(MockBehavior.Strict);
-            var controller = new GroupsController(service.Object, Mapper.Instance);
+            Mock<IGroupService> service = new Mock<IGroupService>(MockBehavior.Strict);
+            GroupsController controller = new GroupsController(service.Object, Mapper.Instance);
 
+            IActionResult result = await controller.Put(1, null);
 
-            var result = controller.Put(1, null) as BadRequestResult;
+            BadRequestResult resultAsBadRequest = result as BadRequestResult;
```
Comment (efueng) [inline SecretSanta/test/SecretSanta.Api.Tests/Controllers/GroupControllerTests.cs]: I would use `var result = controller.Put(1, null) as BadRequestResult;` here instead of breaking it into two statements. Not trying to beat a dead horse, just wanted to mark all these for you in case you decide to change things.

File: SecretSanta/test/SecretSanta.Api.Tests/Controllers/GroupControllerTests.cs [hunk 7]
```diff
@@ -49,130 +51,156 @@ public void GetAllGroups_ReturnsGroups()
...
-                    Id = 2,
-                    Name = group.Name
-                })
-                .Verifiable();
 
-            var controller = new GroupsController(service.Object, Mapper.Instance);
+            Mock<IGroupService> service = new Mock<IGroupService>();
 
-            var result = controller.Put(2, group) as NoContentResult;
+            service.Setup( x => x.GetById(2) )
```
Comment (efueng) [inline SecretSanta/test/SecretSanta.Api.Tests/Controllers/GroupControllerTests.cs]: Consider using ReturnsAsync so you can avoid calling `Task.Run()`

File: SecretSanta/test/SecretSanta.Api.Tests/Controllers/GroupControllerTests.cs [hunk 8]
```diff
@@ -49,130 +51,156 @@ public void GetAllGroups_ReturnsGroups()
...
+                                                                        Id = 2,
+                                                                        Name = group.Name
+                                                                    }
+                                                              )
+                                                     ).Verifiable();
 
-            Assert.IsNotNull(result);
+            GroupsController controller = new GroupsController(service.Object, Mapper.Instance);
+
+            IActionResult result = await controller.Put(2, group);
```
Comment (efueng) [inline SecretSanta/test/SecretSanta.Api.Tests/Controllers/GroupControllerTests.cs]: Again, I'd combine these two statements into one:
`var result = await controller.Put(2, group) as NoContentResult;`

File: SecretSanta/test/SecretSanta.Api.Tests/Controllers/GroupControllerTests.cs [hunk 9]
```diff
@@ -49,130 +51,156 @@ public void GetAllGroups_ReturnsGroups()
...
+            Mock<IGroupService> service = new Mock<IGroupService>(MockBehavior.Strict);
+            GroupsController controller = new GroupsController(service.Object, Mapper.Instance);
 
-            var result = controller.Delete(groupId);
+            IActionResult result = await controller.Delete(groupId);
 
-            Assert.IsTrue(result is BadRequestObjectResult);
+            bool isResultBadRequestObj = result is BadRequestObjectResult;
+
+            Assert.IsTrue(isResultBadRequestObj);
```
Comment (efueng) [inline SecretSanta/test/SecretSanta.Api.Tests/Controllers/GroupControllerTests.cs]: You're basically doing a boolean check on a boolean here. When you do `bool isResultBadRequestObj = result is BadRequestObjectResult;` you're basically saying `thisBool gets thatBool`, and then you do an Assert on `thisBool`. You can shorten this to `Assert.IsTrue(result is BadRequestObjectResult);`

File: SecretSanta/test/SecretSanta.Api.Tests/Controllers/GroupControllerTests.cs [hunk 10]
```diff
@@ -49,130 +51,156 @@ public void GetAllGroups_ReturnsGroups()
...
 
         [TestMethod]
-        public void DeleteGroup_ReturnsNotFoundWhenTheGroupFailsToDelete()
+        public async Task DeleteGroup_ReturnsNotFoundWhenTheGroupFailsToDelete()
         {
-            var service = new Mock<IGroupService>();
+            Mock<IGroupService> service = new Mock<IGroupService>();
             service.Setup(x => x.DeleteGroup(2))
-                .Returns(false)
+                .Returns(Task.Run(() => false))
```
Comment (efueng) [inline SecretSanta/test/SecretSanta.Api.Tests/Controllers/GroupControllerTests.cs]: Using `ReturnsAsync` lets you avoid calling `Task.Run()`

File: SecretSanta/test/SecretSanta.Api.Tests/Controllers/GroupControllerTests.cs [hunk 11]
```diff
@@ -49,130 +51,156 @@ public void GetAllGroups_ReturnsGroups()
...
                 .Verifiable();
             var controller = new GroupsController(service.Object, Mapper.Instance);
 
-            var result = controller.Delete(2);
+            IActionResult result = await controller.Delete(2);
+
+            bool isResultNotFound = result is NotFoundResult;
+
+            Assert.IsTrue(isResultNotFound);
```
Comment (efueng) [inline SecretSanta/test/SecretSanta.Api.Tests/Controllers/GroupControllerTests.cs]: Consider condensing lines 162 and 164 back into `Assert.IsTrue(result is NotFoundResult);`

File: SecretSanta/test/SecretSanta.Api.Tests/Controllers/GroupControllerTests.cs [hunk 12]
```diff
@@ -49,130 +51,156 @@ public void GetAllGroups_ReturnsGroups()
...
         [TestMethod]
-        public void DeleteGroup_ReturnsOkWhenGroupIsDeleted()
+        public async Task DeleteGroup_ReturnsOkWhenGroupIsDeleted()
         {
-            var service = new Mock<IGroupService>();
+            Mock<IGroupService> service = new Mock<IGroupService>();
+
             service.Setup(x => x.DeleteGroup(2))
-                .Returns(true)
+                .Returns(Task.Run(() => true))
```
Comment (efueng) [inline SecretSanta/test/SecretSanta.Api.Tests/Controllers/GroupControllerTests.cs]: Consider using `ReturnsAsync`

File: SecretSanta/test/SecretSanta.Api.Tests/Controllers/GroupControllerTests.cs [hunk 13]
```diff
@@ -49,130 +51,156 @@ public void GetAllGroups_ReturnsGroups()
...
                 .Verifiable();
-            var controller = new GroupsController(service.Object, Mapper.Instance);
 
-            var result = controller.Delete(2);
+            GroupsController controller = new GroupsController(service.Object, Mapper.Instance);
+
+            IActionResult result = await controller.Delete(2);
+
+            bool isResultOk = result is OkResult;
```
Comment (efueng) [inline SecretSanta/test/SecretSanta.Api.Tests/Controllers/GroupControllerTests.cs]: Consider condensing lines 182 and 184 into `Assert.IsTrue(result is OkResult);`

File: SecretSanta/test/SecretSanta.Api.Tests/Controllers/GroupControllerTests.cs [hunk 14]
```diff
@@ -49,130 +51,156 @@ public void GetAllGroups_ReturnsGroups()
...
         }
 
         private static void AssertAreEqual(GroupViewModel expected, Group actual)
         {
-            if (expected == null && actual == null) return;
-            if (expected == null || actual == null) Assert.Fail();
-
-            Assert.AreEqual(expected.Id, actual.Id);
-            Assert.AreEqual(expected.Name, actual.Name);
+            if (expected == null && actual == null)
```
Comment (efueng) [inline SecretSanta/test/SecretSanta.Api.Tests/Controllers/GroupControllerTests.cs]: You can get rid of the else blocks here, because if the two conditionals evaluate as false, the final two asserts are still reachable. If either of the conditionals evaluate as true, then you return out of this method anyways.

File: SecretSanta/src/SecretSanta.Domain/Services/PairingService.cs [hunk 15]
```diff
@@ -0,0 +1,124 @@
...
+            {
+                List<Pairing> pairings = new List<Pairing>();
+                int originalListSize = userIds.Count;
+
+                List<int> randomOrderUserId = new List<int>();
+
+                //lock (_LockKey); //may not be necessary, but placing it for thread safety in incase of multiple thread reaching the Random.next
+                while (randomOrderUserId.Count < originalListSize)
+                {
+                    Monitor.Enter(_lock); //Alternative way I tried to lock. Trying to use lock (_lockKey), I thought didn't quite work when test
```
Comment (Keboo) [inline SecretSanta/src/SecretSanta.Domain/Services/PairingService.cs]: Not sure why `lock` would not work here. Typically you want to avoid using the `Monitor` class directly.

File: SecretSanta/test/SecretSanta.Domain.Tests/Services/PairingServicesTests.cs [hunk 16]
```diff
@@ -0,0 +1,246 @@
...
+
+                Assert.AreEqual(userCount, randomPairings.Count);
+                Assert.IsTrue(IsValidPairings(randomPairings));
+
+                List<Pairing> pairingsFound = await pairingService.GetPairingsList(DefaultGroup.Id);
+                Assert.AreEqual(userCount, pairingsFound.Count);
+            }
+        }
+
+        private bool IsValidPairings(List<Pairing> randomPairings)
```
Comment (Keboo) [inline SecretSanta/test/SecretSanta.Domain.Tests/Services/PairingServicesTests.cs]: Very nice checks
